### DISCUSSION TITLE
Why is `poetry run` required when inside a conda?

### DISCUSSION THREAD

**Discussion by mil-ad** [python-poetry/poetry#7478] [Q&A]
Why do I have add the `poetry run` prefix when I'm using a conda environment? It's extremely clunky and doing yet another `poetry shell` within the conda environment is very inelegant. I thought `poetry install` from a lock file is basically just a wrapper around `pip install` so don't understand why `poetry run` is required.

**Comment by neersighted** [comment]
It is entirely unclear what workflow you are using. `poetry install` is not a simple pip wrapper, but I don't think that matters here.

Poetry will use the conda environment if it is active, instead of managing its own. That means it will install into the environment already in use in your shell, and no special steps are required to use that Python environment. You'll have to be a lot more specific for us to understand what challenge you are facing.

**Comment by mil-ad** [comment]
> Poetry will use the conda environment if it is active, instead of managing its own. That means it will install into the environment already in use in your shell, and no special steps are required to use that Python environment.

Interesting. I'll give it another ago and try to come up with a simple reproducible bug but basically if I run any of my scripts (inside a conda env) with `poetry run python blah.py` they work and if I simply do `python blah.py` none of the dependencies are found.

**Reply by Secrus** [reply]
That's expected. `poetry run xxx` is basically `activate venv; run xxx; deactivate venv`

**Reply by mil-ad** [reply]
that contradicts what @neersighted said:

> That means it will install into the environment already in use in your shell, and no special steps are required to use that Python environment.

what exactly does poetry detect when running inside a conda env then?

**Reply by finswimmer** [reply]
Poetry will create and activate a venv if it detects, that there is no venv currently activated. This detection works by looking at the the environment variables `VIRTUAL_ENV` and `CONDA_PREFIX`. If one of them has a value, and in case of `CONDA_PREFIX` is set, the conda env name is not "base", Poetry will not activate or create a new env. (https://github.com/python-poetry/poetry/blob/3ed013392cafeab1341c08599d1e22d5186bd4b9/src/poetry/utils/env.py#L716)

**Reply by mil-ad** [reply]
Thanks for the clarification. I'm still not sure why a `poetry run` prefix is needed if poetry actually knows that it's running inside a conda environment. It's very clunky.
